File: Graph/KosarajuSCC.java
package Graph;

/**
 * Created by sunkai on 2017/9/5
 */
public class KosarajuSCC {
    //搜索图中的强连通分量
    private boolean[] marked;
    private int[] id;
    private int count;

    public KosarajuSCC(Diagraph G) {
        marked = new boolean[G.V()];
        id = new int[G.V()];
        DepthFIrstOrder order = new DepthFIrstOrder(G.reverse());
        for (int s : order.reversePost()) {
            if (!marked[s]) {
                dfs(G, s);
                count++;
            }
        }
    }

    private void dfs(Diagraph G, int v) {
        marked[v] = true;
        id[v] = count;
        for (int w : G.adj(v)) {
            if (!marked[w]) {
                dfs(G, w);
            }
        }
    }

    public boolean stronglyConnected(int v, int w) {
        return id[v] == id[w];
    }

    public int id(int v) {
        return id[v];
    }

    public int count() {
        return count;
    }
}

File: Algorithm_Exercises/Algorithm (4th edition)/Algorithm/Chapter4/DiGraph/KosarajuSCC.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Chapter4.DiGraph
{
    public class KosarajuSCC : ISCC
    {
        public IDiGraph G { get; }

        private readonly bool[] marked;
        private readonly int[] id; //强连通分量标识符
        public int Count { get; private set; }

        public KosarajuSCC(IDiGraph G)
        {
            this.G = G;
            marked = new bool[G.V];
            id = new int[G.V];
            DepthFirstOrder order = new DepthFirstOrder(G.Reverse());
            foreach(int s in order.ReversePost)
                if(!marked[s])
                {
                    Dfs(G, s);
                    Count++;
                }
        }

        private void Dfs(IDiGraph g, int s)
        {
            marked[s] = true;
            id[s] = Count;
            foreach (int w in G.Adj(s))
            {
                if (!marked[w])
                    Dfs(g, w);
            }
        }

        public int Id(int v) => id[v];

        public bool StronglyConnected(int v, int w) => id[v] == id[w];
    }
}
